FSharpLu.Json/Compact.fs

```diff
@@ -25,67 +63,67 @@ open ConverterHelpers
+             writer.WriteEndObject()
+        | Some(v) when v.GetType().GetTypeInfo().IsGenericType && v.GetType().GetGenericTypeDefinition() = typedefof<Option<_>> ->
+            writer.WriteStartObject()
+            writer.WritePropertyName(convertName SomeFieldIdentifier)
+            serializer.Serialize(writer, v)
+            writer.WriteEndObject()             
+        | Some(v) -> serializer.Serialize(writer, v)
+
+    override this.WriteJson(writer:JsonWriter, value:obj, serializer:JsonSerializer) =
         let t = value.GetType()
+        
+        let convertName =
+            match serializer.ContractResolver with
+            | :? CamelCasePropertyNamesContractResolver -> toCamel
+            | _ -> id
+
         // Option type?
         if isOptionType t then
-            let cases = FSharpType.GetUnionCases(t)
-            let none, some = cases.[0], cases.[1]
-
-            let case, fields = FSharpValue.GetUnionFields(value, t)
-
-            if case = none then
-                () // None is serialized as just null
-
-            // Some _
-            else
-                // Handle cases `Some None` and `Some null`
-                let innerType = some.GetFields().[0].PropertyType
-                let innerValue = fields.[0]
-                if isNull innerValue then
-                    writer.WriteStartObject()
-                    writer.WritePropertyName(convertName SomeFieldIdentifier)
-                    writer.WriteNull()
-                    writer.WriteEndObject()
-                // Some v with v <> null && v <> None
-                else
-                    // Is it nesting another option: `(e.g., "Some (Some ... Some ( ... )))"`
-                    // or any other type with a field named 'Some'?
-                    if hasFieldNamedSome innerType then
-                        // Preserved the nested structure through boxing
-                        writer.WriteStartObject()
-                        writer.WritePropertyName(convertName SomeFieldIdentifier)
-                        serializer.Serialize(writer, innerValue)
-                        writer.WriteEndObject()
-                    else
-                        // Type is option<'a> where 'a does not have a field named 'Some
-                        // (and therfore in particular is NOT an option type itself)
-                        // => we can simplify the Json by omitting the `Some` boxing
-                        // and serializing the nested object directly
-                        serializer.Serialize(writer, innerValue)
-
+            memorise
+                (fun (t: Type) ->                     
+                    let mi = this.GetType().GetTypeInfo().GetMethod("WriteOptionType")
+                    let gm = mi.MakeGenericMethod(t.GetTypeInfo().GetGenericArguments().[0])
+                    let d = gm.CreateDelegate(typeof<Action<JsonWriter, JsonSerializer, obj>>) :?> Action<JsonWriter, JsonSerializer, obj>
```

Comment:
Is it worth the overhead of creating a .net delegate for every single serialized option type, rather than just directly parsing the option type `t` (as before)?

Reply:
Good question - I was seeing overhead here with the existing logic and doing this seemed to remove it as a significant mention in my profiler. Will parse it as before expect with the memorised helpers I did to see if there is any performance difference. On a side note using the existing option parsing logic seems to fix the tests so that's why the build is failing.